sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+                    : _stream.Read(buffer, offset, count);
+                offset += copiedBytes;
+                count -= copiedBytes;
+                _position += copiedBytes;
+                totalCopiedBytes += copiedBytes;
+
+                // We've run out of bytes in the current block.
+                if (copiedBytes == 0)
+                {
+                    // We hit the end of the blob with the last download call.
+                    if (_position == _length)
+                    {
+                        return totalCopiedBytes;
+                    }
+
+                    // Download the next block
+                    else
+                    {
+                        await Download(async, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            while (count > 0);
+            return totalCopiedBytes;
+        }
+
+        private async Task Download(bool async, CancellationToken cancellationToken)
+        {
+            Response<T> response;
+
+            HttpRange range = new HttpRange(_position, _bufferSize);
+
+            try
+            {
+                response = async
+                    ? await _downloadToAsyncFunc(range, _requestConditions, default, cancellationToken).ConfigureAwait(false)
+                    : _downloadToFunc(range, _requestConditions, default, cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
+                throw;
+            }
+
+            Stream networkStream = (Stream)typeof(T).GetProperty("Content").GetValue(response.Value, null);
+
+            _stream.SetLength(0);
+
+            if (async)
+            {
+                await networkStream.CopyToAsync(
+                    _stream,
+                    Constants.DefaultBufferSize,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                networkStream.CopyTo(
+                    _stream,
+                    Constants.DefaultBufferSize);
+            }
+
+            _stream.Position = 0;
+            _lastDownloadBytes = response.GetRawResponse().Headers.ContentLength.GetValueOrDefault();
+            _length = GetBlobLength(response);
+        }
+
+        private static void ValidateReadParameters(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException($"{nameof(buffer)}", $"{nameof(buffer)} cannot be null.");
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} cannot be less than 0.");
+            }
+
+            if (offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} cannot exceed {nameof(buffer)} length.");
```

Comment:
Why?  We can always give them less data than they're asking for on a Read call.

Reply:
I think it's reasonable to validated the that customer is giving us an offset that is inbounds for the buffer.  Let's say buffer .Length is 1024, and the customer hands me an offset of 2048.  What are we supposed to do here?  We can't write any bytes.

Was this comment about `offset + count > buffer.Length`?

Reply:
I removed the `offset + count > buffer.Length` check in the next iteration, and added logic in `ReadInternal()` to handle offset + count > buffer.Length, assuming offset is inbounds.

Reply:
I think I meant to put this comment on the `offset + count > buffer.Length` line because that's the one where we can still give them some data, even if it's less than `count`.